File: src/commands/list.ts
/**
 * List command - Display local issues with interactive selector
 */

import { readdirSync } from 'node:fs';
import { join } from 'node:path';
import { select } from '@inquirer/prompts';
import chalk from 'chalk';
import ora from 'ora';
import { readConfig } from '../lib/config.js';
import { getProjectMobiusPath, readParentSpec } from '../lib/local-state.js';
import { resolvePaths } from '../lib/paths.js';
import { syncBackendStatuses } from '../lib/status-sync.js';
import type { Backend } from '../types.js';

export interface ListOptions {
  backend?: Backend;
}

/**
 * List all local issues and present an interactive selector.
 * Outputs the selected issue identifier to stdout.
 */
export async function list(options: ListOptions = {}): Promise<void> {
  const paths = resolvePaths();
  const config = readConfig(paths.configPath);
  const backend = options.backend ?? config.backend;

  // Sync statuses from backend before displaying
  const syncSpinner = ora({
    text: 'Syncing statuses from backend...',
    stream: process.stderr,
    color: 'blue',
  }).start();

  const syncResult = await syncBackendStatuses(backend);

  if (syncResult.synced > 0 || syncResult.failed > 0) {
    const parts: string[] = [];
    if (syncResult.synced > 0) parts.push(`${syncResult.synced} synced`);
    if (syncResult.failed > 0) parts.push(`${syncResult.failed} unreachable`);
    syncSpinner.succeed(`Status sync: ${parts.join(', ')}`);
  } else {
    syncSpinner.succeed('Status sync complete.');
  }

  const issuesPath = join(getProjectMobiusPath(), 'issues');

  let entries: { name: string; isDirectory: () => boolean }[];
  try {
    entries = readdirSync(issuesPath, { withFileTypes: true });
  } catch {
    console.error(chalk.yellow('No local issues found.'));
    console.error(chalk.gray('Run `mobius refine <issue-id>` to create local issue state.'));
    return;
  }

  const dirs = entries.filter((e) => e.isDirectory());

  if (dirs.length === 0) {
    console.error(chalk.yellow('No local issues found.'));
    console.error(chalk.gray('Run `mobius refine <issue-id>` to create local issue state.'));
    return;
  }

  const choices: Array<{ name: string; value: string }> = [];

  for (const dir of dirs) {
    const issueId = dir.name;
    const spec = readParentSpec(issueId);
    if (!spec) continue;

    const statusColor =
      spec.status === 'Done'
        ? chalk.green
        : spec.status === 'In Progress'
          ? chalk.cyan
          : chalk.gray;

    choices.push({
      name: `${chalk.bold(spec.identifier)}  ${spec.title}  ${statusColor(`[${spec.status}]`)}`,
      value: spec.identifier,
    });
  }

  if (choices.length === 0) {
    console.error(chalk.yellow('No valid local issues found.'));
    console.error(chalk.gray('Issue directories exist but parent specs could not be read.'));
    return;
  }

  const selected = await select(
    { message: 'Select an issue:', choices },
    { output: process.stderr }
  );

  console.log(selected);
}

File: rust/mobius/src/commands/list.rs
//! List command - Display local issues with interactive selector

use colored::Colorize;
use std::fs;

use crate::config::loader::read_config;
use crate::config::paths::resolve_paths;
use crate::local_state::{get_project_mobius_path, read_parent_spec};
use crate::types::enums::Backend;

pub fn run(backend_override: Option<&str>) -> anyhow::Result<()> {
    let paths = resolve_paths();
    let config = read_config(&paths.config_path).unwrap_or_default();
    let _backend: Backend = if let Some(b) = backend_override {
        b.parse().unwrap_or(config.backend)
    } else {
        config.backend
    };

    let issues_path = get_project_mobius_path().join("issues");

    let entries = match fs::read_dir(&issues_path) {
        Ok(entries) => entries,
        Err(_) => {
            eprintln!("{}", "No local issues found.".yellow());
            eprintln!(
                "{}",
                "Run `mobius refine <issue-id>` to create local issue state.".dimmed()
            );
            return Ok(());
        }
    };

    let mut dirs: Vec<String> = Vec::new();
    for entry in entries.flatten() {
        if entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            if let Some(name) = entry.file_name().to_str() {
                dirs.push(name.to_string());
            }
        }
    }

    if dirs.is_empty() {
        eprintln!("{}", "No local issues found.".yellow());
        eprintln!(
            "{}",
            "Run `mobius refine <issue-id>` to create local issue state.".dimmed()
        );
        return Ok(());
    }

    dirs.sort();

    let mut choices: Vec<(String, String)> = Vec::new();

    for issue_id in &dirs {
        if let Some(spec) = read_parent_spec(issue_id) {
            let status_color = match spec.status.as_str() {
                "Done" => spec.status.green().to_string(),
                "In Progress" => spec.status.cyan().to_string(),
                _ => spec.status.dimmed().to_string(),
            };

            let display = format!(
                "{}  {}  [{}]",
                spec.identifier.bold(),
                spec.title,
                status_color
            );
            choices.push((display, spec.identifier));
        }
    }

    if choices.is_empty() {
        eprintln!("{}", "No valid local issues found.".yellow());
        eprintln!(
            "{}",
            "Issue directories exist but parent specs could not be read.".dimmed()
        );
        return Ok(());
    }

    let items: Vec<&str> = choices.iter().map(|(display, _)| display.as_str()).collect();

    let selection = dialoguer::Select::new()
        .with_prompt("Select an issue")
        .items(&items)
        .interact()?;

    // Output selected issue identifier to stdout
    println!("{}", choices[selection].1);

    Ok(())
}
